File: src/ast.rs
use std::rc::Rc;

use derive_more::{Display, IsVariant};
use enum_as_inner::EnumAsInner;

use crate::SourceSpan;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, IsVariant, EnumAsInner)]
pub enum Ast {
    File(Rc<AstFile>),
    Procedure(Rc<AstProcedure>),
    ProcedureType(Rc<AstProcedureType>),
    Parameter(Rc<AstParameter>),
    Scope(Rc<AstScope>),
    LetDeclaration(Rc<AstLet>),
    VarDeclaration(Rc<AstVar>),
    Name(Rc<AstName>),
    Integer(Rc<AstInteger>),
    Call(Rc<AstCall>),
    Return(Rc<AstReturn>),
    Unary(Rc<AstUnary>),
    Binary(Rc<AstBinary>),
    If(Rc<AstIf>),
    While(Rc<AstWhile>),
    Cast(Rc<AstCast>),
    Assign(Rc<AstAssign>),
}

impl Ast {
    pub fn get_location(&self) -> SourceSpan {
        match self {
            Ast::File(ast) => ast.location.clone(),
            Ast::Procedure(ast) => ast.location.clone(),
            Ast::ProcedureType(ast) => ast.location.clone(),
            Ast::Parameter(ast) => ast.location.clone(),
            Ast::Scope(ast) => ast.location.clone(),
            Ast::LetDeclaration(ast) => ast.location.clone(),
            Ast::VarDeclaration(ast) => ast.location.clone(),
            Ast::Name(ast) => ast.location.clone(),
            Ast::Integer(ast) => ast.location.clone(),
            Ast::Call(ast) => ast.location.clone(),
            Ast::Return(ast) => ast.location.clone(),
            Ast::Unary(ast) => ast.location.clone(),
            Ast::Binary(ast) => ast.location.clone(),
            Ast::If(ast) => ast.location.clone(),
            Ast::While(ast) => ast.location.clone(),
            Ast::Cast(ast) => ast.location.clone(),
            Ast::Assign(ast) => ast.location.clone(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstFile {
    pub location: SourceSpan,
    pub expressions: Vec<Ast>,
}

#[derive(Clone, Debug, Display, PartialEq, Eq, PartialOrd, Ord, EnumAsInner)]
pub enum CallingConvention {
    #[display(fmt = "#cdecl")]
    CDecl,
    #[display(fmt = "#stdcall")]
    StdCall,
    #[display(fmt = "#fastcall")]
    FastCall,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstProcedure {
    pub location: SourceSpan,
    pub name: String,
    pub parameters: Vec<Rc<AstParameter>>,
    pub return_type: Ast,
    pub calling_convention: CallingConvention,
    pub body: AstProcedureBody,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstProcedureType {
    pub location: SourceSpan,
    pub parameter_types: Vec<Ast>,
    pub calling_convention: CallingConvention,
    pub return_type: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstParameter {
    pub location: SourceSpan,
    pub mutable: bool,
    pub name: String,
    pub typ: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, IsVariant, EnumAsInner)]
pub enum AstProcedureBody {
    ExternName(String),
    Scope(Rc<AstScope>),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstScope {
    pub location: SourceSpan,
    pub expressions: Vec<Ast>,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstLet {
    pub location: SourceSpan,
    pub name: String,
    pub typ: Option<Ast>,
    pub value: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstVar {
    pub location: SourceSpan,
    pub name: String,
    pub typ: Option<Ast>,
    pub value: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstName {
    pub location: SourceSpan,
    pub name: String,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstInteger {
    pub location: SourceSpan,
    pub value: u128,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstCall {
    pub location: SourceSpan,
    pub operand: Ast,
    pub arguments: Vec<Ast>,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstReturn {
    pub location: SourceSpan,
    pub value: Option<Ast>,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, EnumAsInner)]
pub enum UnaryOperator {
    Identity,
    Negation,
    LogicalNot,
    PointerType,
    AddressOf,
    Dereference,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstUnary {
    pub location: SourceSpan,
    pub operator: UnaryOperator,
    pub operand: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, EnumAsInner)]
pub enum BinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Remainder,
    Equal,
    NotEqual,
    LessThan,
    GreaterThan,
    LessThanEqual,
    GreaterThanEqual,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstBinary {
    pub location: SourceSpan,
    pub left: Ast,
    pub operator: BinaryOperator,
    pub right: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstIf {
    pub location: SourceSpan,
    pub condition: Ast,
    pub then_expression: Ast,
    pub else_expression: Option<Ast>,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstWhile {
    pub location: SourceSpan,
    pub condition: Ast,
    pub then_expression: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstCast {
    pub location: SourceSpan,
    pub typ: Ast,
    pub operand: Ast,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, EnumAsInner)]
pub enum AstAssignDirection {
    Left,
    Right,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct AstAssign {
    pub location: SourceSpan,
    pub direction: AstAssignDirection,
    pub operand: Ast,
    pub value: Ast,
}

File: src/Ast.cpp
#include "Ast.hpp"

#include <iostream>

namespace Langite {

    void AstSearcher::Visit(AstFile& ast) {
        for (auto& expression : ast.Expressions)
            expression->Accept(*this);
    }

    void AstSearcher::Visit(AstBlock& ast) {
        for (auto& expression : ast.Expressions)
            expression->Accept(*this);
    }

    void AstSearcher::Visit(AstUnary& ast) {
        ast.Operand->Accept(*this);
    }

    void AstSearcher::Visit(AstBinary& ast) {
        ast.Left->Accept(*this);
        ast.Right->Accept(*this);
    }

    void AstSearcher::Visit(AstFieldAccess& ast) {
        ast.Operand->Accept(*this);
    }

    void AstSearcher::Visit(AstIndex& ast) {
        ast.Operand->Accept(*this);
        ast.Indexer->Accept(*this);
    }

    void AstSearcher::Visit(AstCall& ast) {
        ast.Operand->Accept(*this);
        for (auto& argument : ast.Arguments)
            argument->Accept(*this);
    }

    void AstSearcher::Visit(AstGenericInstantiation& ast) {
        ast.Operand->Accept(*this);
        for (auto& genericArgument : ast.GenericArguments)
            genericArgument->Accept(*this);
    }

    void AstSearcher::Visit(AstParenthesisedExpression& ast) {
        ast.Expression->Accept(*this);
    }

    void AstSearcher::Visit(AstDeclaration& ast) {
        ast.Type->Accept(*this);
    }

    void AstSearcher::Visit(AstConstDeclaration& ast) {
        if (ast.GenericParameters)
            for (auto& genericParameter : *ast.GenericParameters)
                genericParameter->Accept(*this);
        if (ast.Type)
            (*ast.Type)->Accept(*this);
        ast.Value->Accept(*this);
    }

    void AstSearcher::Visit(AstName&) {}

    void AstSearcher::Visit(AstWildcard&) {}

    void AstSearcher::Visit(AstInteger&) {}

    void AstSearcher::Visit(AstFloat&) {}

    void AstSearcher::Visit(AstString&) {}

    void AstSearcher::Visit(AstFunction& ast) {
        for (auto& parameter : ast.Parameters)
            parameter->Accept(*this);
        ast.ReturnType->Accept(*this);
        if (ast.Body)
            (*ast.Body)->Accept(*this);
    }

    void AstSearcher::Visit(AstProcedure& ast) {
        for (auto& parameter : ast.Parameters)
            parameter->Accept(*this);
        ast.ReturnType->Accept(*this);
        if (ast.Body)
            (*ast.Body)->Accept(*this);
    }

    void AstSearcher::Visit(AstReturn& ast) {
        if (ast.Value)
            (*ast.Value)->Accept(*this);
    }

    void AstSearcher::Visit(AstIf& ast) {
        ast.Condition->Accept(*this);
        ast.ThenBlock->Accept(*this);
        if (ast.ElseScope)
            (*ast.ElseScope)->Accept(*this);
    }

    void AstSearcher::Visit(AstBuiltin&) {}

    class AstDumper: public AstVisitor {
        size_t Indent = 0;

        void PrintIndent() {
            for (size_t i = 0; i < Indent; i++) {
                std::cout << "    ";
            }
        }
    public:
        virtual void Visit(AstFile& ast) override {
            PrintIndent();
            std::cout << "- File" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Expressions:" << std::endl;
            Indent++;
            for (auto& expression : ast.Expressions) {
                expression->Accept(*this);
            }
            Indent--;
            Indent--;
        }

        virtual void Visit(AstBlock& ast) override {
            PrintIndent();
            std::cout << "- Block" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Expressions:" << std::endl;
            Indent++;
            for (auto& expression : ast.Expressions) {
                expression->Accept(*this);
            }
            Indent--;
            Indent--;
        }

        virtual void Visit(AstUnary& ast) override {
            PrintIndent();
            std::cout << "- Unary" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "OperatorToken: '" << TokenKind_ToString(ast.OperatorToken.Kind) << '\''
                      << std::endl;
            PrintIndent();
            std::cout << "Operand:" << std::endl;
            Indent++;
            ast.Operand->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstBinary& ast) override {
            PrintIndent();
            std::cout << "- Binary" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "OperatorToken: '" << TokenKind_ToString(ast.OperatorToken.Kind) << '\''
                      << std::endl;
            PrintIndent();
            std::cout << "Left:" << std::endl;
            Indent++;
            ast.Left->Accept(*this);
            Indent--;
            PrintIndent();
            std::cout << "Right:" << std::endl;
            Indent++;
            ast.Right->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstFieldAccess& ast) override {
            PrintIndent();
            std::cout << "- Field Access" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Field Name: " << std::get<std::string_view>(ast.FieldNameToken.Data)
                      << std::endl;
            PrintIndent();
            std::cout << "Operand:" << std::endl;
            Indent++;
            ast.Operand->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstIndex& ast) override {
            PrintIndent();
            std::cout << "- Index" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Operand:" << std::endl;
            Indent++;
            ast.Operand->Accept(*this);
            Indent--;
            PrintIndent();
            std::cout << "Indexer:" << std::endl;
            Indent++;
            ast.Indexer->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstCall& ast) override {
            PrintIndent();
            std::cout << "- Call" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Operand:" << std::endl;
            Indent++;
            ast.Operand->Accept(*this);
            Indent--;
            PrintIndent();
            std::cout << "Arguments:" << std::endl;
            Indent++;
            for (auto& argument : ast.Arguments) {
                argument->Accept(*this);
            }
            Indent--;
            Indent--;
        }

        virtual void Visit(AstGenericInstantiation& ast) override {
            PrintIndent();
            std::cout << "- Generic Instantiation" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Operand:" << std::endl;
            Indent++;
            ast.Operand->Accept(*this);
            Indent--;
            PrintIndent();
            std::cout << "Generic Arguments:" << std::endl;
            Indent++;
            for (auto& argument : ast.GenericArguments) {
                argument->Accept(*this);
            }
            Indent--;
            Indent--;
        }

        virtual void Visit(AstParenthesisedExpression& ast) override {
            PrintIndent();
            std::cout << "- Parenthesised Expression" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Expression:" << std::endl;
            Indent++;
            ast.Expression->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstDeclaration& ast) override {
            PrintIndent();
            std::cout << "- Declaration: " << static_cast<void*>(&ast) << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Name: '" << std::get<std::string_view>(ast.NameToken.Data) << '\''
                      << std::endl;
            PrintIndent();
            std::cout << "Type:" << std::endl;
            Indent++;
            ast.Type->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstConstDeclaration& ast) override {
            PrintIndent();
            std::cout << "- Const Declaration: " << static_cast<void*>(&ast) << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Name: '" << std::get<std::string_view>(ast.NameToken.Data) << '\''
                      << std::endl;
            if (ast.GenericParameters) {
                PrintIndent();
                std::cout << "Generic Parameters:" << std::endl;
                Indent++;
                for (auto& parameter : *ast.GenericParameters) {
                    parameter->Accept(*this);
                }
                Indent--;
            }
            if (ast.Type) {
                PrintIndent();
                std::cout << "Type:" << std::endl;
                Indent++;
                (*ast.Type)->Accept(*this);
                Indent--;
            }
            PrintIndent();
            std::cout << "Value:" << std::endl;
            Indent++;
            ast.Value->Accept(*this);
            Indent--;
            Indent--;
        }

        virtual void Visit(AstName& ast) override {
            PrintIndent();
            std::cout << "- Name: '" << std::get<std::string_view>(ast.NameToken.Data) << '\''
                      << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Resolved Declaration: " << ast.ResolvedDeclaration << std::endl;
            Indent--;
        }

        virtual void Visit(AstWildcard&) override {
            PrintIndent();
            std::cout << "- Wildcard" << std::endl;
        }

        virtual void Visit(AstInteger& ast) override {
            PrintIndent();
            std::cout << "- Integer: " << std::get<size_t>(ast.IntegerToken.Data) << std::endl;
        }

        virtual void Visit(AstFloat& ast) override {
            PrintIndent();
            std::cout << "- Float: " << std::get<double>(ast.FloatToken.Data) << std::endl;
        }

        virtual void Visit(AstString& ast) override {
            PrintIndent();
            auto& string = std::get<std::vector<char>>(ast.StringToken.Data);
            std::cout << "- String: \"" << std::string_view{ string.data(), string.size() } << '"'
                      << std::endl;
        }

        virtual void Visit(AstFunction& ast) override {
            PrintIndent();
            std::cout << "- Function" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Parameters:" << std::endl;
            Indent++;
            for (auto& parameter : ast.Parameters) {
                parameter->Accept(*this);
            }
            Indent--;
            PrintIndent();
            std::cout << "Return Type:" << std::endl;
            Indent++;
            ast.ReturnType->Accept(*this);
            Indent--;
            if (ast.Body) {
                PrintIndent();
                std::cout << "Body:" << std::endl;
                Indent++;
                (*ast.Body)->Accept(*this);
                Indent--;
            }
            Indent--;
        }

        virtual void Visit(AstProcedure& ast) override {
            PrintIndent();
            std::cout << "- Procedure" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Parameters:" << std::endl;
            Indent++;
            for (auto& parameter : ast.Parameters) {
                parameter->Accept(*this);
            }
            Indent--;
            PrintIndent();
            std::cout << "Return Type:" << std::endl;
            Indent++;
            ast.ReturnType->Accept(*this);
            Indent--;
            if (ast.Body) {
                PrintIndent();
                std::cout << "Body:" << std::endl;
                Indent++;
                (*ast.Body)->Accept(*this);
                Indent--;
            }
            Indent--;
        }

        virtual void Visit(AstReturn& ast) override {
            PrintIndent();
            std::cout << "- Return" << std::endl;
            Indent++;
            if (ast.Value) {
                PrintIndent();
                std::cout << "Value:" << std::endl;
                Indent++;
                (*ast.Value)->Accept(*this);
                Indent--;
            }
            Indent--;
        }

        virtual void Visit(AstIf& ast) override {
            PrintIndent();
            std::cout << "- If" << std::endl;
            Indent++;
            PrintIndent();
            std::cout << "Condition:" << std::endl;
            Indent++;
            ast.Condition->Accept(*this);
            Indent--;
            PrintIndent();
            std::cout << "Then Block:" << std::endl;
            Indent++;
            ast.ThenBlock->Accept(*this);
            Indent--;
            if (ast.ElseScope) {
                PrintIndent();
                std::cout << "Else Scope:" << std::endl;
                Indent++;
                (*ast.ElseScope)->Accept(*this);
                Indent--;
            }
            Indent--;
        }

        virtual void Visit(AstBuiltin& ast) override {
            PrintIndent();
            std::cout << "- Builtin" << std::endl;
            Indent++;
            PrintIndent();
            auto& string = std::get<std::vector<char>>(ast.StringToken.Data);
            std::cout << "String: \"" << std::string_view{ string.data(), string.size() } << '"'
                      << std::endl;
            Indent--;
        }
    };

    void DumpAst(Ast& ast) {
        AstDumper dumper;
        ast.Accept(dumper);
    }

}
